### ISSUE TITLE
`use proc_macro` regressed from nightly-2018-09-14 to nightly 2018-09-15 in edition 2018

### ISSUE THREAD

**Issue by LukasKalbertodt** [rust-lang/rust#54418]
Use statements for `proc_macro` that worked in nightly-(2ab3eba30 2018-09-14) don't work in (e4ba1d41e 2018-09-15) (including all newer nightlies up to at least (3bc2ca7e4 2018-09-20) and the RC1 beta) anymore.  

I tried this code:

**src/lib.rs**
```rust
use proc_macro::TokenStream;

#[proc_macro]
pub fn foo(input: TokenStream) -> TokenStream {
    input
}
```

**Cargo.toml**
```toml
[package]
name = "proc-fail"
version = "0.1.0"
authors = ["me"]
edition = "2018"

[lib]
proc-macro = true

[dependencies]
```

As said, compiling with (2ab3eba30 2018-09-14), everything works. But compiling with (e4ba1d41e 2018-09-15) results in:

**Error (1)**
```
error[E0432]: unresolved import `proc_macro`
 --> src/lib.rs:3:5
  |
3 | use proc_macro::TokenStream;
  |     ^^^^^^^^^^ Could not find `proc_macro` in `{{root}}`
```

---

I tried different configurations with the two different nightly versions. The error in `edition = "2015"` without the extern crate is expected, of course.

**With nightly-(2ab3eba30 2018-09-14):**

| | `edition = "2015"` | `edition = "2018"` |
| - | - | - |
| with `extern crate proc_macro` | works | works |
| without `extern crate ...` | fails (3) | works |

**With nightly- (e4ba1d41e 2018-09-15):**

| | `edition = "2015"` | `edition = "2018"` |
| - | - | - |
| with `extern crate proc_macro` | works | fails (2) |
| without `extern crate ...` | fails (3) | fails (1) |

**Error (2)**
```
error[E0432]: unresolved import `proc_macro`
 --> src/lib.rs:3:5
  |
3 | use proc_macro::TokenStream;
  |     ^^^^^^^^^^ Did you mean `self::proc_macro`?
```

**Error (3)**
```
error[E0432]: unresolved import `proc_macro`
 --> src/lib.rs:3:5
  |
3 | use proc_macro::TokenStream;
  |     ^^^^^^^^^^ Did you mean `registrar::proc_macro`?
```

---

I couldn't really find any information about this. So the only way to make it work right now is to either use edition 2015 with extern crate or edition 2018 with extern crate and `use self::proc_macro`.

Is this intended or an unwanted regression?

**Comment by petrochenkov** [comment]
@LukasKalbertodt 
>Is this intended or an unwanted regression?

Intended, https://github.com/rust-lang/rust/pull/54116 is the relevant PR.

**Comment by LukasKalbertodt** [comment]
@petrochenkov Thanks!

So what's the intended solution for Rust 2018 now? `extern crate proc_macro; use self::proc_macro`? That seems wrong :/
As I understand it, `meta` is only a placeholder and not yet usable.

**Comment by petrochenkov** [comment]
Yes, `extern crate proc_macro;` is the solution on stable, until `--extern crate_name` is stabilized in some form.

**Comment by LukasKalbertodt** [comment]
@petrochenkov But just adding `extern crate proc_macro;` doesn't solve the problem. Here as a full example again:

**src/lib.rs**
```rust
extern crate proc_macro;

use proc_macro::TokenStream;


#[proc_macro]
pub fn foo(input: TokenStream) -> TokenStream {
    input
}
```

**Cargo.toml**
```toml
[package]
name = "proc-fail"
version = "0.1.0"
authors = ["me"]
edition = "2018"

[lib]
proc-macro = true

[dependencies]
```

Executing `cargo +nightly-2018-09-16 build` results in (note `nightly-2018-09-16` is `(e4ba1d41e 2018-09-15)`):

```
error[E0432]: unresolved import `proc_macro`
 --> src/lib.rs:3:5
  |
3 | use proc_macro::TokenStream;
  |     ^^^^^^^^^^ Did you mean `self::proc_macro`?
```

Is this also intended?

**Comment by petrochenkov** [comment]
@LukasKalbertodt 
No, this is not intended.

Minimal reproduction:
```rust
// --edition 2018
extern crate proc_macro;
use proc_macro::TokenStream; // ERROR unresolved import `proc_macro`

fn main() {}
```
cc @eddyb

**Comment by petrochenkov** [comment]
Workaround:
```rust
// --edition 2018
extern crate proc_macro;
use crate::proc_macro::TokenStream; // ERROR unresolved import `proc_macro`

fn main() {}
```

**Comment by eddyb** [comment]
We can't make `extern crate` extend the extern prelude, can we? I haven't seen a plan for doing so, or how it would be sound wrt fixpoint searches.

**Comment by petrochenkov** [comment]
@eddyb 
The question is why `use proc_macro::...` doesn't resolve to `proc_macro` from the current module (`self::proc_macro`) introduced by the extern crate item.

**Comment by petrochenkov** [comment]
Oh, crap.
`feature(uniform_path)` is still not enabled by default.

**Comment by petrochenkov** [comment]
So, yes, with "anchored path" model (https://github.com/rust-lang/rust/issues/53130) this works as intended, even if it looks pretty bad.
